### PR TITLE
Do not substitute beneath arithmetic terms in the non-linear solver

### LINKED ISSUE #3307: Assertion failure NRA formula
Hi,
for the following NRA formula, CVC4 gives an assertion failure

```
(set-logic NRA)
(declare-fun a () Real)
(declare-fun b () Real)
(assert
 (and
  (> b 1)
  (< a 0)
  (>= (/ 0 (+ (* a b) (/ (- a) 0))) a)
  )
 )
(check-sat)
```

```
(error "Assertion failure.
bool CVC4::theory::arith::NonlinearExtension::simpleCheckModelMsum(const std::map<CVC4::NodeTemplate<true>, CVC4::NodeTemplate<true> >&, bool)
/home/windomin/solver_binaries/cache/CVC4/src/theory/arith/nonlinear_extension.cpp:1741:

  false
")

```
OS:Ubuntu 18.04
Revision: 7988675

### PR DESCRIPTION
**Pull request by ajreynol**
This fixes #3307 and #3472. 

The bug was caused by substituting within arithmetic terms that appear in non-linear monomials. This means that a term like `f(x)*f(x) + x*x` would turn into `f(7)*f(7) + 7*7` if the non-linear extension inferred that `x = 7`.  The problem was then that we have cached information for the term `f(x)` but not `f(7)`.  The solution is to not traverse beneath applications of functions that belong to a theory other than arithmetic for certain substitution operations.

Note that the bad behavior was hindering our ability to say "sat", although it is not related to correctness.

### CODE REVIEW

**Comment by ajreynol** [comment]
Btw, I'm not super happy about adding a special substitution function in this PR; I considered generalizing this to a generic utility function.  This is up for discussion.

**Comment by ahmed-irfan** [comment]
cool.... checking it now

File: src/theory/arith/nonlinear_extension.cpp [hunk 1]
```diff
@@ -387,6 +387,85 @@ Node NonlinearExtension::computeModelValue(Node n, unsigned index) {
...
+      std::vector<Node> children;
+      if (cur.getMetaKind() == kind::metakind::PARAMETERIZED)
+      {
+        children.push_back(cur.getOperator());
+      }
+      for (const Node& cn : cur)
+      {
+        it = visited.find(cn);
+        Assert(it != visited.end());
+        Assert(!it->second.isNull());
```
Comment (mpreiner) [inline src/theory/arith/nonlinear_extension.cpp]: Use `CVC4_DCHECK`

File: src/theory/arith/nonlinear_extension.cpp [hunk 2]
```diff
@@ -387,6 +387,85 @@ Node NonlinearExtension::computeModelValue(Node n, unsigned index) {
...
+      }
+      if (childChanged)
+      {
+        ret = nm->mkNode(cur.getKind(), children);
+      }
+      visited[cur] = ret;
+    }
+  } while (!visit.empty());
+  Assert(visited.find(n) != visited.end());
+  Assert(!visited.find(n)->second.isNull());
```
Comment (mpreiner) [inline src/theory/arith/nonlinear_extension.cpp]: Ditto

**Comment by ajreynol** [comment]
We decided to generalize the `arithSubstitute` in this PR to a general utility with callbacks.

**Comment by ajreynol** [comment]
After some more thought about this, I think I actually am happy with the status of this PR. The downside of generalizing and moving arithSubstitute to expr/node_algorithm.h is that we may be overdesigning.  I think we should only generalize this when there are at least 2 compelling variants of substitute that fit this pattern, but after some thought I'm not sure there is even another use case in the existing code. 

@ahmed-irfan @4tXJ7f @mpreiner what do you think?

**Comment by ahmed-irfan** [comment]
there is a conflict in `test/regress/CMakeLists.txt`

**Pull request merged**